clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -106,13 +107,21 @@ public void configure(TransportLayer transportLayer, PrincipalBuilder principalB
 
             setSaslState(handshakeRequestEnable ? SaslState.SEND_HANDSHAKE_REQUEST : SaslState.INITIAL);
 
-            // determine client principal from subject.
-            if (!subject.getPrincipals().isEmpty()) {
-                Principal clientPrincipal = subject.getPrincipals().iterator().next();
-                this.clientPrincipalName = clientPrincipal.getName();
-            } else {
-                clientPrincipalName = null;
+            // determine client principal from subject for Kerberos. For other mechanisms, the authenticated
+            // principal (username for PLAIN and SCRAM) is used as authorization id.
+            if (mechanism.equals(SaslConfigs.GSSAPI_MECHANISM)) {
+                try {
+                    Principal clientPrincipal = subject.getPrincipals().iterator().next();
```

Comment:
@ijuma If `getPrincipals` is not empty, I dont think the code would throw any exceptions. But `next()` can throw an exception regardless of whether we check `isEmpty()` first since the principals are cleared on a different thread. So rather than  process the failure of `isEmpty()` as well as `next()`, the PR expects `next()` to succeed for `GSSAPI`. If not it indicates that a refresh is in progress and we fail this connection. It is not ideal, but it avoids synchronization (we dont want to block the network thread). It also doesn't make sense to continue with a null `clientPrincipal` in some scenarios for GSSAPI.